io/slow5/slow5.go

```diff
@@ -0,0 +1,440 @@
+slow5 files contain both general metadata about the sequencing run and raw
+signal reads from the sequencing run. This raw signal can be used directly or
+basecalled and used for alignment.
+
+More information on slow5 can be found here: https://github.com/hasindu2008/slow5tools
+*/
+package slow5
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+/******************************************************************************
+Oct 10, 2021
+
+slow5 parser begins here. Specification below:
+https://hasindu2008.github.io/slow5specs/slow5-v0.1.0.pdf
+
+slow5 is able to combine multiple sequencing runs into a single file format,
+but we read each sequencing run separately. Each sequencing run contains a
+header with metadata and a list of reads. However, unlike many other file
+formats, a slow5 file should almost never be read into a common struct, since
+most runs are large, and will take a ton of memory. Instead, the default way
+to parse slow5 files is to produce a list of headers and a channel of raw
+reads. In order to connect the two (if needed), use ReadGroupID.
+
+Nanopore changes the attributes found in the header quite often, so we store
+most of these attributes in a map for future proofing. Even the binary file
+format, blow5, does not have types for these attributes, and just stores them
+as a long string.
+
+Reads have 8 required columns, and a few auxillary. These are typed, since they
+are what will probably be used in real software.
+
+Cheers mate,
+
+Keoni
+
+******************************************************************************/
+
+// Header contains metadata about the sequencing run in general.
+type Header struct {
+	ReadGroupID        uint32
+	Slow5Version       string
+	Attributes         map[string]string
+	EndReasonHeaderMap map[string]int
+}
+
+// Read contains metadata and raw signal strengths for a single nanopore read.
+type Read struct {
+	ReadID       string
+	ReadGroupID  uint32
+	Digitisation float64
+	Offset       float64
+	Range        float64
+	SamplingRate float64
+	LenRawSignal uint64
+	RawSignal    []int16
+
+	// Auxiliary fields
+	ChannelNumber string
+	MedianBefore  float64
+	ReadNumber    int32
+	StartMux      uint8
+	StartTime     uint64
+	EndReason     string // enum{unknown,partial,mux_change,unblock_mux_change,data_service_unblock_mux_change,signal_positive,signal_negative}
+
+	Error error // in case there is an error while parsing!
+}
+
+var knownEndReasons = map[string]bool{"unknown": true,
+	"partial":                         true,
+	"mux_change":                      true,
+	"unblock_mux_change":              true,
+	"data_service_unblock_mux_change": true,
+	"signal_positive":                 true,
+	"signal_negative":                 true,
+}
+
+// Parser is a flexible parser that provides ample
+// control over reading slow5 sequences.
+// It is initialized with NewParser.
+type Parser struct {
+	// reader keeps state of current reader.
+	reader       bufio.Reader
```

Comment:
should these capitalized and exported?

Reply:
Idk, should they? I don't have a strong opinion either way.

Reply:
I think safest path is to leave them unexported, this protects the underlying implementation allowing it to change if we find a better way of parsing. Like Gophers say: You can always export them in the future if it comes to that. But unexporting them is a breaking change.